### ISSUE TITLE
TE copies for fully assembled repeats

### ISSUE THREAD

**Issue by zhaolina0827** [clemgoub/dnaPipeTE#51]
Hello！
I am going to calculate TE copies for fully assembled repeats, but I don’t know which file n, N, L come from?
Best wishes!
Lina Zhao

**Comment by zhaolina0827** [comment]
Hello！
I am going to calculate TE copies for fully assembled repeats, but I don’t know which file n, N, L come from?
Best wishes!
Lina Zhao

**Comment by clemgoub** [comment]
Hello!

Here is the details:

n = count of read mapping one TE family (you can get it from "reads_per_component_and_annotation" 
N = total number or reads in the blast sample. You can know this by doing
`grep -c '>' renamed.blasting_reads.fasta`
L = size of TE family consensus

In the paper, I first cluster the dnaPipeTE contigs (from the file `Trinity.fasta`) with CD-HIT-EST. CD-HIT gives one representative sequence per cluster (the longest). I used the length of this "representative" sequence as `L` and the sum of the reads counts from all the dnaPipeTE contings belonging to the same cluster as `n`.

Let me know if you have further questions!

Cheers,

Clément

**Comment by r-mashoodh** [comment]
Hi Clément, 

Could I take the same approach to comparing copy numbers across different samples (from the same species)?

1) run dnaPipeTE on 2 different samples
2) compute copy number as described above for each sample separately
3) then merge and compare

?

Thanks again!

**Comment by clemgoub** [comment]
Dear @r-mashoodh,

I indeed recommend the procedure shown in the dnaPipeTE [paper](https://academic.oup.com/gbe/article/7/4/1192/533768) as well as [this one](https://pubmed.ncbi.nlm.nih.gov/32041215/):
1. Rename each Trinity.fasta adding the sample name to the contig
2. Merge the samples to compare and run CD-HIT-EST as described in the papers
3. Identify shared repeats
4. For each cluster in each species, sum the counts of the contigs clustered together, not only the longest 

Let me know if you need further help with that!

Cheers,

Clément

**Comment by r-mashoodh** [comment]
Hi @clemgoub !

Thanks so much for this. I clustered according to the paper ... some clarification questions:
1) some clusters don't have contigs from all the samples .. is this expected? so would this mean 0 for that sample or do you only keep clusters in which all samples are represented?
2) is L = size of TE family consensus, still going to be the longest contig in the cluster (regardless from which cluster it came from?)
3) how do you decide that copy numbers might be *different* between samples?

cheers and thanks again for all your help!

**Comment by clemgoub** [comment]
Hello @r-mashoodh, these a good questions! Meaning my answers are not absolute!

> 1. some clusters don't have contigs from all the samples .. is this expected? so would this mean 0 for that sample or do you only keep clusters in which all samples are represented?

Counts of 0 can be either because this particular TE family is absent, or that is at a too low frequency for dnaPipeTE to find enough reads to assemble. Thus I would only compare clusters shared by the different samples to avoid this bias. This is still possible of course that a TE family is absent from one sample, but if you suspect this I recommend looking a it manually before including the 0 counts.

> 2. is L = size of TE family consensus, still going to be the longest contig in the cluster (regardless from which cluster it came from?)

This is a little bit tricky. The longest sequence is often the most complete, though there can be some artifact/chimeras, making one sequence the longest. In clustering of different samples: the longest will become the reference, but it ultimately comes only from one sample. If the samples are the same species, I think this is fine.

> 3. how do you decide that copy numbers might be _different_ between samples?

This is the limit of the method: because a lower copy sample will have fewer reads, the assembled consensus can be more fragmented than for a sample with more copy and thus a better consensus. Thus picking the "reference" sequence for both samples might be tricky as described before. Maybe you could generate a new consensus from each cluster (sort of average of each cluster) and use it as the new reference to estimate the copy numbers. However, I would check a liberal amount of these reconstituted sequences by hand to see if they represent well the TE. You can use [TE-Aid](https://github.com/clemgoub/TE-Aid) to check your consensus.

**Comment by r-mashoodh** [comment]
hi @clemgoub!

So I made an attempt combine contigs and re-cluster as described in your paper.

Specifically:
1. I renamed each Trinity.fasta file so that contigs could be traced back to the original sample
2. combined all the Trinity.fasta files
3. ran cd-hit in your paper (`cd-hit-est -i Combined.Trinity.fasta -o Combined.Cluster.fasta -aS 0.8 -c 0.8 -G 0 -g 1 -T 32`)

So now what I want to do is count reads mapping to each cluster and compare. However, I find that clusters have different TE class/family assignments

Some clusters have several different family assignments ... here's an example:

<img width="455" alt="clstr8" src="https://user-images.githubusercontent.com/17982017/135750296-17b1be34-ad12-43ec-a022-9a3804beaee7.png">

And this varies from sample to sample ... a little bit more agreement for cluster 8 in another sample ...

<img width="450" alt="clustr8sampl2" src="https://user-images.githubusercontent.com/17982017/135750387-e1b23f09-d8a7-4a53-8f1d-ff16d0103019.png">

I realise there may not be a straightforward answer, but how does one decide what's what?

Thanks in advance!

r.
